Hibernate Reactive extension

aguibert opened the pull request:
For issue #8883 

This gets Hibernate Reactive pulled in as a Quarkus extension that will work in JVM mode an Native mode.

Example configuration:
```
quarkus.datasource.db-kind=postgresql
quarkus.datasource.username=hibernate_orm_test
quarkus.datasource.password=hibernate_orm_test

quarkus.datasource.reactive=true
quarkus.datasource.reactive.url=${postgres.reactive.url}

quarkus.datasource.jdbc=false
```

Example usage:
```java
import org.hibernate.reactive.mutiny.Mutiny;
import org.hibernate.reactive.stage.Stage;
import io.smallrye.mutiny.Uni;
// ...

@Inject
Stage.Session session;

@GET
@Path("/reactiveFind")
@Produces(MediaType.APPLICATION_JSON)
public Uni<GuineaPig> reactiveFind() {
    return Uni.createFrom().completionStage(session.find(GuineaPig.class, 5));
}
```
Full user-facing code can be seen here: https://github.com/quarkusio/quarkus/pull/9060/files#diff-0037a9a64055493b35d0b9f7e5e58d54

### Major items to note for now:
- User needs to enable 2 Quarkus extensions: `quarkus-hibernate-reactive`, the appropriate reactive driver (e.g. `quarkus-reactive-pg-client`)
- User may now `@Inject` `Stage.SessionFactory` and `Stage.Session`
- No integration with Mutiny for now, but once that design firms up over in Hibernate Reactive we can consider something different.

### Limitations (planning to solve after this PR)
- Applications cannot use traditional Hibernate ORM at the same time as Hibernate Reactive
- (internal) Refactor common pieces of the ORM extension into a common dependency and pull that in. Idea being that the Reactive extension does not pull in the ORM extension
- Only the default reactive datasource is supported. (This limitation is shared with ORM only supporting the default jdbc datasource as well)

### Items that still need to be addressed:
 - ~Schema gen support in HibernateRX would eliminate the need for JDBC dependency/config~
 - ~I need to look into native mode support for this test~
 - ~Refactor `quarkus-reactive-datasource` to depend on the generic vertx-sql-client API so it produces a generic `PoolBuildItem`~
 - Remove the need for configuration via persistence.xml

aguibert committed:
Initial Hibernate Reactive extension

aguibert committed:
App booting with Hibernate Reactive

aguibert committed:
Initial find test case passing

aguibert committed:
Support for injecting RxSession

aguibert committed:
Add support for injecting Mutiny.Session

aguibert committed:
Make Rx producers @Typed so no qualifier is needed

aguibert committed:
Add VertxPoolBuildItem for depending on generic Vertx Pools

aguibert committed:
Modifications for native mode compatibility

aguibert committed:
Make RxSession and RxSessionFactory producers @DefaultBean so app may override them

Sanne commented:
Hi @aguibert , great progress!

Quick question; the JDBC drivers are necessary just for schema generation / validation, correct?

Sanne commented:
It strikes me that your example is configured using a `persistence.xml` file.  FYI I've been supporting that format mostly so to make life easier for people migrating to Quarkus, but I'm not sure if it should be around long term - primarily as it's not a good idea to set the expectation to users that they can use all same properties they are used to.

For a new project like this one, I think supporting exclusively the properties from the standard Quarkus configuration file would be better; I might be wrong but it's easier to add more configuration flavours later.

Sanne commented:
@FroMage may I ask you to play around with how it feels as a reactive user?

Also, feel free to start exploring Panache RX on top of this?

aguibert committed:
Remove direct dependency on Mutiny

aguibert commented:
@Sanne @FroMage I added some commits just now which gets everything running properly in native mode too and did some additional code cleanup, so this PR is ready to be reviewed in more detail now. I'm still going to leave it as a draft since there is no release version of HibernateRX though, so we can't merge until we have that.

## Performance/footprint

Overall, I'm pretty happy with the initial numbers compared to ORM:
| Feature | Native image size | Native boot time |
| -------- | ------------------- | ---|
| Hibernate ORM | 58MB | 70 - 75ms |
| Hibernate RX | 63MB | 50 - 60ms |

Note:
- Both test apps also enabled `resteasy-mutiny` and `resteasy-jsonb`
- I haven't done any time to first response or throughput measurements yet
- RX image size is larger (as expected) because it builds on top of the `quarkus-hibernate-orm` extension and also requires both JDBC and reactive (Vertx) drivers to be in place.

## Mutiny integration

For now I've removed the `Mutiny.Session` producer which eliminates this extensions dependency on Mutiny, so apps can only inject `RxSession` and `RxProducer`, but can easily write a producer for `Mutiny.Session` if they want.

A few reasons for that:
1. Mutiny integration in HibernateRX is still being discussed/prototyped (see https://github.com/hibernate/hibernate-rx/issues/94)
2. `Mutiny.Session` is a concrete class with a 1-args ctor which exposed a bug in Arc that needs to be fixed (see https://github.com/quarkusio/quarkus/issues/9132)

We can always add the Mutiny integration later, perhaps with a `quarkus-hibernate-rx-mutiny` integration extension (similar to what we have for `quarkus-resteasy-jsonb` and `quarkus-resteasy-mutiny`)

aguibert commented:
also @Sanne was the comment "feel free to start exploring Panache RX too" directed at Stephane or me?

FroMage commented:
That's for me, I believe. I need to rebase my Panache RX branch on Hibernate RX. Which will probably be hard because it supported transactions, queries, reactive relations, and didn't depend on ORM.

Either I need to drop all those features/tests, or I start from scratch. Not sure yet what is easiest. But it will be a branch based on this branch, which means every rebase will affect my branch.

> there is no released version of HibernateRX though

They should get a CR out the door ASAP to make our life easier.

Sanne commented:
> also @Sanne was the comment "feel free to start exploring Panache RX too" directed at Stephane or me?

right that's meant for Stephane. Sorry it wasn't clear: FYI he already has many pretty advanced experiments for Panache RX, just that the previous ones are based on existing ORM.

Sanne commented:
> Not sure yet what is easiest. But it will be a branch based on this branch, which means every rebase will affect my branch.

Sure, no pressure :) It might not be too bad to rebase though as this is going to be a separate extension?

> They should get a CR out the door ASAP to make our life easier.

Not sure about it deserving a `CR` qualifier already, but we can certainly start looking for preview tag releases.

aguibert commented:
@Sanne regarding the persistence.xml comment:

> It strikes me that your example is configured using a persistence.xml file. FYI I've been supporting that format mostly so to make life easier for people migrating to Quarkus, but I'm not sure if it should be around long term - primarily as it's not a good idea to set the expectation to users that they can use all same properties they are used to.

Do you think it would be good to follow suit from what HibernateRX (standalone) allows? It would be odd to a user if they could use a persistence.xml in standalone, but not in Quarkus.

Sanne commented:
> Do you think it would be good to follow suit from what HibernateRX (standalone) allows? It would be odd to a user if they could use a persistence.xml in standalone, but not in Quarkus.

No I don't think that would be odd. Many other projects when running "in Quarkus" are not configured as they normally are - we have a mantra of "single configuration file" in Quarkus that all extension writers are asked to follow.

The fact that the Hibernate ORM extension is allowing the use of `persistence.xml` as a configuration source is an oddity, and it might be temporary. Technically I'd wish to deprecate it but many configuration options have no corresnponding in the Quarkus configuration so we'd first need to polish the "new" configurability.

With a new extension such as this one we might want to do it right from the beginning - we can always reintroduce it later if this strategy changes.

Besides - not really relevant - but while `Hibernate Reactive` (name is changing!) in standalone mode can be configured via a persistence.xml today, I'm not sure if that's the right thing going forward. We'll discuss that on the relevant project though.

aguibert committed:
Rebrand from HibernateRX to Hibernate Reactive

aguibert committed:
Update Hibernate Reactive extension to not require JDBC resources

aguibert commented:
@Sanne @gavinking I've rebased on Gavin's schema generation work and got that working in Quarkus now so that JDBC configuration/dependencies is no longer required. This reduces the native-image footprint from 63MB --> 50MB and the boot time from 50-60ms --> 25-30ms 

Next I'll look at removing the need for (and blocking) usage of persistence.xml

Sanne commented:
that's great!

aguibert committed:
Remaining rebranding from RX --> Reactive

gavinking commented:
@aguibert note that the new Mutiny API is now available [here](https://github.com/hibernate/hibernate-reactive/pull/149).

aguibert committed:
Support for Hibernate Reactive Mutiny API

FroMage commented:
@aguibert Do the tests pass for you?

I've rebased the extension on `master` locally, to try to get moving with Panache Reactive, but I can't seem to run my tests, due to ArC issues, and when I run the hibernate-reactive IT, I get this:

```
2020-06-03 16:37:36,763 WARN  [io.qua.dep.QuarkusAugmentor] (main) Using Java versions older than 11 to build Quarkus applications is deprecated and will be disallowed in a future release!
2020-06-03 16:37:37,247 WARN  [io.qua.agr.dep.AgroalProcessor] (build-22) The Agroal dependency is present but no JDBC datasources have been defined.
2020-06-03 16:37:39,471 WARN  [org.hib.eng.jdb.con.int.ConnectionProviderInitiator] (main) HHH000181: No appropriate connection provider encountered, assuming application will be supplying connections
2020-06-03 16:39:39,758 INFO  [io.quarkus] (main) Quarkus 999-SNAPSHOT on JVM started in 123.134s. Listening on: http://0.0.0.0:8081
2020-06-03 16:39:39,759 INFO  [io.quarkus] (main) Profile test activated. 
2020-06-03 16:39:39,759 INFO  [io.quarkus] (main) Installed features: [agroal, cdi, hibernate-orm, hibernate-reactive, mutiny, narayana-jta, reactive-pg-client, resteasy, resteasy-jsonb, resteasy-mutiny, vertx]
2020-06-03 16:40:40,791 ERROR [org.jbo.res.res.i18n] (vert.x-eventloop-thread-0) RESTEASY002020: Unhandled asynchronous exception, sending back 500: io.netty.channel.ConnectTimeoutException: connection timed out: /192.168.1.22:5431
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe$1.run(AbstractNioChannel.java:261)
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:170)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
```

FroMage commented:
Perhaps it's a docker issue. For my own tests I run postgres locally and I get issues like these:

```
	[error]: Build step io.quarkus.hibernate.orm.deployment.HibernateOrmProcessor#configurationDescriptorBuilding threw an exception: io.quarkus.deployment.configuration.ConfigurationError: Hibernate extension cannot guess the dialect as no database kind is specified by 'quarkus.datasource.db-kind'
```

When it's really set:

```
quarkus.datasource.db-kind=postgresql
quarkus.datasource.username=test
quarkus.datasource.password=test
quarkus.datasource.reactive.url=vertx-reactive:postgresql://localhost:5432/test
quarkus.datasource.jdbc=false
```

Or:

```
Caused by: javax.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type javax.persistence.EntityManagerFactory and qualifiers [@Default]
	- java member: io.quarkus.hibernate.reactive.runtime.ReactiveSessionFactoryProducer#emf
	- declared on CLASS bean [types=[io.quarkus.hibernate.reactive.runtime.ReactiveSessionFactoryProducer, java.lang.Object], qualifiers=[@Default, @Any], target=io.quarkus.hibernate.reactive.runtime.ReactiveSessionFactoryProducer]
	at io.quarkus.arc.processor.Beans.resolveInjectionPoint(Beans.java:486)
	at io.quarkus.arc.processor.BeanInfo.init(BeanInfo.java:362)
	at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:234)
```

FroMage commented:
Looks like `quarkus-hibernate-reactive-deployment` depends on `quarkus-hibernate-orm-deployment` which is configured with `JdbcDataSourceBuildItem` which is only produced if `JDBC` is enabled, so I can't actually disable JDBC or else I get this error.

FroMage commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java:
```diff
@@ -166,6 +166,7 @@ public SystemPropertyBuildItem enforceDisableRuntimeEnhancer() {
     }
 
     @BuildStep
+    //<<<<<<< HEAD
```
Looks like a leftover.

FroMage commented on extensions/hibernate-reactive/runtime/src/main/java/io/quarkus/hibernate/reactive/runtime/ReactiveSessionProducer.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    public void disposeMutinySession(@Disposes Uni<Mutiny.Session> reactiveSession) {
+    	// TODO: @AGG this is not working, need to check w/ Clement to figure out the proper way
```
I doubt the version for `CompletionStage` works either: this creates a new `Uni`/`CompletionStage` that must be used in order for it to be terminated. This has no side-effect.

aguibert commented:
@FroMage it certainly was working last time I pushed it, but it may need another manual update if master has changed enough. I know @Sanne was doing some refactoring in the ORM space and he asked that I freeze this branch for now so he can incorporate it in his refactoring.
I'll rebase on master locally and see what seems to have changed

FroMage commented:
Adding the JDBC driver and config allowed me to go further, so it looks like it's required on master.

aguibert commented:
FYI I just double checked locally and the tests passed with the current branch as-is (which has `quarkus.datasource.jdbc=false`) so something must have changed in master. I checked w/ Sanne and he said he hasn't delivered his ORM refactors yet

aguibert commented:
@FroMage I just got done rebasing on the Quarkus master and HR master, I had to make quite a few changes to the extension code, but the test code was able to remain the same (i.e. JDBC still disabled). I'll push these rebase fixes once I get the OK from @Sanne to un-freeze

aguibert commented:
Actually, now that HR 1.0.0.Alpha5 is on maven central, I could make this into a non-draft PR and we could actually merge this in. WDYT @Sanne ?

Sanne committed:
Adapt to Hibernate Reactive 1.0.0.Alpha5 changes in SPI

Sanne committed:
Hibernate Reactive now suppporgs schema generation as well

Sanne committed:
Add some design notes for maintenance

Sanne commented:
@aguibert I've already rebased this and updated to Alpha5; indeed I don't plan to do much more work but I have some minor polishing locally now which I'd prefer to keep.

I'll share it soon :  +1 to expedite merging, we can do more work in collaboration as soon as it's in.

Before merging it though, I have some doubts that I need to clarify in the area of "what will it break in existing apps" e.g. all integration with Hibernate Envers seems disabled as you commented out its registration.

Sanne committed:
Re-enable Hibernate Envers

Sanne committed:
Review runtime dependency tree

Sanne committed:
Code formatting rules applied

Sanne committed:
Simplify inconsistencies in choice of PersistenceProvider implementation

Sanne committed:
Customize the reactive service initiators list

Sanne committed:
Use testing properties connecting to localhost by default

Sanne committed:
Adding a log cleanup filter for Hibernate Reactive

Sanne commented:
@aguibert I think I've done enough :) Trying to stick to very minor improvements

Could you take my branch `hibernate-rx` and push-force to overwrite this one? So we could reuse the PR and not loose its comments & al?

You'll need to use --force as I rebased and had some conflicts to solve.

Other than conflicts and having switch the Quarkus version, your commits are unaffected - there's some more from me but they should be straight forward for you to review.

aguibert commented:
Thanks @Sanne ! I've updated my PR with your commits and the rebase now. 

w.r.t. integration aspects such as Hibernate Envvers, I wanted this PR to just focus on getting core hibernate reactive, so that myself (or others) can iterate on the core work to add the support for non-core things like envvers. Unfortunately I underestimated the amount of time this would have to be in draft form.

Moving this PR out of draft now though!

Sanne commented:
great! Right, assuming CI is feeling better we should merge it, then we can continue further optimisations in smaller pieces, especially to allow more people to collaborate.

I forgot to check one aspect - the CI job. Will we need to enlist it in one of the actions, and possibly override some of those database urls?

Sanne commented:
ah, just noticed:
```
extensions/hibernate-reactive/runtime/src/main/resources/META-INF/quarkus-extension.yaml
```
shoudln't be flagged as "stable" :)

aguibert commented:
Good catch @Sanne, I'll mark it as non-stable and also enlist in CI

Sanne commented on .github/workflows/ci-actions.yml:
```diff
@@ -122,7 +122,7 @@ jobs:
     services:
       keycloak:
-        image: quay.io/keycloak/keycloak:10.0.1
+        image: quay.io/keycloak/keycloak:10.0.0
```
how does this relate?

There's another change to keycloak version below.

aguibert replied:
good catch, must have been a bad auto-merge when I rebased -- will fix this

aguibert committed:
Mark Hibernate Reactive extension as preview

aguibert committed:
Add hibernate reactive tests to CI

Sanne reviewed (approved):
@aguibert great, thanks. I was about to add this other comment - no longer necessary as you just changed the commit, but let me share the info with you anyway:

(was writing in context of the "--privileged" option in the now-gone changes to the DB2 configuration):
```
I'd suggest to undo this change: I don't know why - but locally I also had trouble with DB2 when I don't use the `privileged` option.

I was assuming it could have been something odd of my own machine setup, but apparently the DB2/reactive tests which are in master now failed when running this CI job.
```

Sanne commented:
I really wish I could merge this now, but CI status doesn't seem reassuring - even if those failures don't seem related.

aguibert commented:
all of the data-related jobs passed, which is promising.

I can't seem to find a way to re-spin individual jobs. For example the `Security3` group failed after 14s due to some connection issue in JDK 11 setup. @gsmet do you have powers to respin some of these for us?

geoand commented:
> all of the data-related jobs passed, which is promising.
> 
> I can't seem to find a way to re-spin individual jobs. For example the `Security3` group failed after 14s due to some connection issue in JDK 11 setup. @gsmet do you have powers to respin some of these for us?

You can restart individual jobs - you need to restart the whole CI run

Sanne commented:
@aguibert we've restarted them all once again.. crossing fingers.

aguibert commented:
@Sanne looks like CI is working again and passed all green. Should be all set to merge this one now

FroMage commented:
Now I know why the IT test passes: you have a `persistence.xml` file. Is that meant to be required, is that temporary or is it an oversight to not have tested configuration via the normal quarkus config?

Sanne commented:
It definitely needs to work without persistence.xml .

I'd even go a step further and not allow `persistence.xml` for Reactive, as
I consider it only a temporary thing for blocking ORM too: we do have it as
the initial requirements predate Quarkus design ideas (the original intent
was the challenge to see how far we could stick to the spec and still have
it all work in native).


On Mon, 8 Jun 2020, 10:38 Stéphane Épardaud, <notifications@github.com>
wrote:

> Now I know why the IT test passes: you have a persistence.xml file. Is
> that meant to be required, is that temporary or is it an oversight to not
> have tested configuration via the normal quarkus config?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/quarkusio/quarkus/pull/9060#issuecomment-640491712>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAKMTNN5B7HSD4IPKPFA6TRVSWQFANCNFSM4MY5PLTQ>
> .
>

aguibert commented:
Yep I need to do a follow-on to remove the persistence.xml and furthermore block it from being used with HR

Sanne commented:
I've created https://github.com/quarkusio/quarkus/issues/9859
